Nullptr crash in elementCannotHaveEndTag

rniwa opened the pull request:
#### a32bbf9a620959bd50deb3c6d58626db4c78c6d4
<pre>
Nullptr crash in elementCannotHaveEndTag
<a href="https://bugs.webkit.org/show_bug.cgi?id=263100">https://bugs.webkit.org/show_bug.cgi?id=263100</a>
&lt;rdar://116331745&gt;

Reviewed by Sihui Liu and Wenson Hsieh.

Added an early return for when current becomes null. Somehow this happens in wild
but we don&apos;t know how we get into that state so keep a debug assert there instead.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):

Canonical link: <a href="https://commits.webkit.org/269320@main">https://commits.webkit.org/269320@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96849aa430712c1f060f3e96a5abe88269458adf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22401 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20550 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22722 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21571 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22427 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19256 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24942 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20112 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26369 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20210 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20338 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24229 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20779 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17683 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20130 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24341 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2776 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20727 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -262,6 +262,9 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
+            ASSERT(current);
+            if (!current)
+                break;
             shouldEmitCloseTag = !(targetNode.document().isHTMLDocument() && elementCannotHaveEndTag(*current));
```
I think the early return should come before calculating both the flags's values:
```
ASSERT(current);
if (!current)
    break;

shouldAppendNode = ...
shouldEmitCloseTag = ...
```

szewai commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -261,6 +261,10 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
                     current = fragment->host();
             }
 
+            ASSERT(current);
```
Is there anything that guarantees `current` won't be null? (I am not very familiar with this code logic so I am just asking since `parentNode()` above seems can return null)

rniwa replied:
In theory, we should never hit a nullptr here because the traversal stays within the subtree given to this function.

szewai replied:
ah, makes sense.

rniwa committed:
Nullptr crash in elementCannotHaveEndTag
https://bugs.webkit.org/show_bug.cgi?id=263100
<rdar://116331745>

Reviewed by Sihui Liu and Wenson Hsieh.

Added an early return for when current becomes null. Somehow this happens in wild
but we don't know how we get into that state so keep a debug assert there instead.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):

Canonical link: https://commits.webkit.org/269320@main
